snappy/snapd_client.go

```diff
@@ -82,3 +94,93 @@ func (a *ClientAdapter) Remove(name string, options *client.SnapOptions) (string
 func (a *ClientAdapter) ServerVersion() (*client.ServerVersion, error) {
 	return a.snapdClient.ServerVersion()
 }
+
+// internal
+func readNTPServer() string {
+	timesyncd, err := ini.Load(timesyncdPath)
+	if err != nil {
+		log.Println("readNTPServer: unable to read /etc/system/timesyncd.conf")
+		return ""
+	}
+
+	section, err := timesyncd.GetSection("Time")
+	if err != nil || !section.HasKey("NTP") {
+		log.Println("readNTPServer: no NTP servers are set")
+		return ""
+	}
+
+	return section.Key("NTP").Strings(" ")[0]
+}
+
+// Write back directly, rather than write-tmp(in same dir)+rename, as per the
+// internals of ini package, although better, write accessible locations may
+// not be on the same filesystem, removing the advantage.
+//  : timeserver-control grants rw access to /etc/systemd/timesyncd.conf
+func saveTimeSyncd(confFile *ini.File) error {
+	timesyncConf, err := os.OpenFile(timesyncdPath, os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		log.Println(fmt.Sprintf("saveTimeSyncd: Failed to open conf file: %v", err))
+		return err
+	}
+
+	defer timesyncConf.Close()
+
+	w := bufio.NewWriter(timesyncConf)
+	_, err = confFile.WriteTo(w)
+	if err != nil {
+		log.Println(fmt.Sprintf("saveTimeSyncd: Failed to write conf file: %v", err))
+		return err
+	}
+	w.Flush()
+	return nil
+}
+
+func writeNTPServer(ntpServer string) {
+	log.Println(fmt.Sprintf("writeNTPServer: server=%s", ntpServer))
+	timesyncd, err := ini.Load(timesyncdPath)
+	if err != nil {
+		log.Println("writeNTPServer: unable to read /etc/system/timesyncd.conf")
+		return
+	}
+
+	section, err := timesyncd.GetSection("Time")
+	if err != nil || !section.HasKey("NTP") {
+		log.Println("writeNTPServer: no NTP servers are set")
+		return
+	}
+
+	ntpKey := section.Key("NTP")
+	servers := ntpKey.Strings(" ")
+	if servers[0] == ntpServer {
+		log.Println("writeNTPServer: NTP server already set")
+		return
+	}
+
+	servers = append([]string{ntpServer}, servers...)
+	ntpKey.SetValue(strings.Join(servers, " "))
+
+	saveTimeSyncd(timesyncd)
+}
+
+// SetCoreConfig sets some aspect of core configuration
+func (a *ClientAdapter) SetCoreConfig(patch map[string]interface{}) (string, error) {
+	for k, v := range patch {
+		fmt.Printf("%s=%v\n", k, v)
```

Comment:
I'm not sure I understand this, this will just print to stdout, is this intended? Or do you actually want something like output := ""; for k,v... { fmt.Sprintf(output, "%s\n%s=%v", output, k,v)}" ?